go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,336 @@
+	observer.Progress(fmt.Sprintf("Veryfing %d empty addresses...", numEmptyAddresses))
+	addresses := generateUnknownAddresses(numEmptyAddresses)
+	for _, addr := range addresses {
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		proof, err := trie.CreateWitnessProof(addr)
+		if err != nil {
+			return err
+		}
+		// expect an empty account
+		if err := verifyAccountProof(rootHash, proof, addr, accountData{}); err != nil {
+			return err
+		}
+	}
+
+	const numKeys = 10
+	observer.Progress(fmt.Sprintf("Veryfing %d empty keys for accounts...", numKeys))
+	for _, addr := range correctAddresses {
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		keys := generateUnknownKeys(numKeys)
+		proof, err := trie.CreateWitnessProof(addr, keys...)
+		if err != nil {
+			return err
+		}
+		if err := verifyStorageProof(rootHash, proof, addr, keys, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// verifyAccountProof verifies the consistency between the input witness proofs and the account.
+func verifyAccountProof(root common.Hash, proof witness.Proof, addr common.Address, data accountData) error {
+	balance, complete, err := proof.GetBalance(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := balance, data.balance; got != want {
+		return fmt.Errorf("balance mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	nonce, complete, err := proof.GetNonce(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := nonce, data.nonce; got != want {
+		return fmt.Errorf("nonce mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	code, complete, err := proof.GetCodeHash(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := code, data.code; got != want {
+		return fmt.Errorf("code mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+
+	return nil
+}
+
+// verifyStorageProof verifies the consistency between the input witness proofs and the storage.
+func verifyStorageProof(root common.Hash, proof witness.Proof, addr common.Address, keys []common.Key, storage map[common.Key]common.Value) error {
+	for _, key := range keys {
+		proofValue, complete, err := proof.GetState(root, addr, key)
+		if err != nil {
+			return err
+		}
+		if !complete {
+			return fmt.Errorf("proof incomplete for address: 0x%x, key: 0x%x", addr, key)
+		}
+		if got, want := proofValue, storage[key]; got != want {
+			return fmt.Errorf("storage mismatch for 0x%x, key 0x%x, got %v, want %v", addr, key, got, want)
+		}
+	}
+
+	return nil
+}
+
+// generateUnknownAddresses generates a slice of addresses that will never appear
+// in real MPT and the witness proof. Each address consists of four random
```

Comment:
"never appear" is not guaranteed by this function. It is just very unlikely.

If you want to make sure that they do not appear, you should pass the trie to check that the address indeed does not exist.

Alternatively, you create `number` random addresses and where you use them, you do not assume that those accounts are empty but you look them up in the MPT. Most of them will be empty, but if by chance you do have an address that is in the MPT, you compare with the actual account state.

Reply:
I see your point, I wanted to save on MPT lookups as an address with 16 trailing zeros is very unlikely in real scenario, but I understand that the verification tool should be rigorous.